### PR TITLE
Supported Convolution and MaxPooling layers

### PR DESCRIPTION
**Pull request by l-bat**

### CODE REVIEW

File: src/runtime/cpu/layers/cpuconvolution.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+#include "runtime/cpu/kernels/kernels.hpp"
+
+deepworks::cpu::CPUConvolution::CPUConvolution(deepworks::LayerInfo&& info)
+    : deepworks::cpu::ICPULayer(std::move(info)),
+      grad_weights(m_info.params()[0].grad()),
+      grad_bias(m_info.params()[1].grad()) {}
+
+void deepworks::cpu::CPUConvolution::validate(const std::vector<deepworks::Tensor>& inputs,
+                                              const std::vector<deepworks::Tensor>& outputs) {
+    DeepWorks_Assert(inputs.size()  == 1u && "Convolutional takes only one input");
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/runtime/cpu/layers/cpuconvolution.cpp:11]: Just not to add new "bad examples" of using `validate`

[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/runtime/cpu/layers/cpuconvolution.cpp:11]: I believe there shouldn't such method at all

File: src/runtime/cpu/layers/cpumaxpooling.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
+#include "runtime/cpu/layers/cpumaxpooling.hpp"
+#include "runtime/cpu/kernels/kernels.hpp"
+
+deepworks::cpu::CPUMaxPooling::CPUMaxPooling(deepworks::LayerInfo&& info)
+    : deepworks::cpu::ICPULayer(std::move(info)) {
+}
+
+void deepworks::cpu::CPUMaxPooling::validate(const std::vector<deepworks::Tensor>& inputs,
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpumaxpooling.cpp:8]: The same, I would remove it

File: src/runtime/cpu/layers/cpumaxpooling.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+void deepworks::cpu::CPUMaxPooling::backward(const std::vector<deepworks::Tensor>& /* inputs */,
+                                             const std::vector<deepworks::Tensor>& /* outputs */,
+                                             const std::vector<deepworks::Tensor>& grad_outputs,
+                                                   std::vector<deepworks::Tensor>& grad_inputs) {
+    const auto& grad_output = grad_outputs.front();
+          auto& grad_input  = grad_inputs.front();
+
+    DeepWorks_Assert(max_indices.shape() == grad_output.shape() && "Call backward without forward");
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpumaxpooling.cpp:50]: In fact it should be done on `backend` lvl, this print is nice for debugging, my concern is that `DeepWorks_Assert` also performs in release.

Comment (TolyaTalamanov) [reply src/runtime/cpu/layers/cpumaxpooling.cpp:50]: I would remove, but up to you

File: src/runtime/cpu/layers/cpumaxpooling.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    auto kernel  = m_info.impl().attrs["kernel"].get<std::array<int, 2>>();
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpumaxpooling.cpp:52]: `const auto&`

File: src/runtime/cpu/layers/cpumaxpooling.hpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#include <deepworks/tensor.hpp>
+#include "runtime/cpu/layers/cpulayer.hpp"
+#include "util/assert.hpp"
+
+namespace deepworks {
+namespace cpu {
+
+class CPUMaxPooling : public ICPULayer {
+public:
+    CPUMaxPooling(deepworks::LayerInfo&& info);
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpumaxpooling.hpp:12]: using

**Review by TolyaTalamanov — APPROVED** [review]
Overall implementation is nice, looking forward for reference to validate it :fire: 

LGTM :+1:

**Comment by apertovs** [comment]
Good job!
But we have sporadic MaxPool1x1 behavior: https://gist.github.com/apertovs/eb58ad7792cdb4836ec6c9ac9f165bc3
Also the test for MaxPool2d 3x3 have strange output: https://gist.github.com/apertovs/b1acd72722a2e1f0013f7c250c89d87e

File: src/runtime/cpu/kernels/kernels.cpp [hunk 6]
```diff
@@ -42,6 +46,202 @@ void deepworks::CPUReLUInputGrad(ConstVector input,
...
+    Matrix grad{grad_output_copy.data(), batch * c, h_out * w_out};
+    auto grad_output_tr = grad.transpose();
+
+    auto out_grad_ptr = grad_output_copy.data();
+    auto indices_ptr  = max_indices.data();
+    auto im2col_ptr   = im2col_buf.data();
+
+    Shape img_shape{1, 1, input_shape[Input::H], input_shape[Input::W]};
+    for (size_t i = 0; i < batch * c; i++) {
+        initializer::zeros(im2col_buf);
```
[the remark this answers is not in the document]
Comment (l-bat) [reply src/runtime/cpu/kernels/kernels.cpp:241]: No

File: src/runtime/cpu/kernels/kernels.cpp [hunk 7]
```diff
@@ -51,6 +55,210 @@ void deepworks::CPUELUInputGrad(ConstVector input, ConstVector grad_output, Vect
...
 
+void deepworks::CPUConvolutionalForward(const Tensor& input,
+                                        const Tensor& weights,
+                                        const Tensor& bias,
+                                        Tensor& output,
+                                        Tensor& im2col_buf,
+                                        const std::array<int, 2>& kernel,
+                                        const std::array<int, 2>& padding,
+                                        const std::array<int, 2>& stride) {
+    auto input_shape = input.shape();
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:66]: `const auto&`

**Review by TolyaTalamanov — APPROVED** [review]

**Pull request merged**